File: core/scripts/Java-Library/src/chh/math/set/OrderedTriple.java
package chh.math.set;

/**
 * An OrderedTriple is an extension (figuratively and literally) on the ordered pair: it has
 * three coordinates rather than two.
 * <p>An OrderedTriple is similar to a {@see Tuple} with 3 components (a 3-tuple),
 * however they are structurally different from a set-theoretic approach:
 * The ordered triple (a,b,c) does not have the same elements as the 3-tuple &lang;a,b,c&rang;
 * and thus they are not equal.</p>
 * <p>{@code OrderedTriple(a,b,c)} is defined as {@code new OrderedPair(new OrderedPair(a,b), c)}.</p>
 * @author  <a href="mailto:chrisharvey2pi@gmail.com">Christopher H. Harvey</a>
 * @version 2013.08.03
 */
public class OrderedTriple extends OrderedPair {
	/** The first coordinate. */
	private final Set abscissa;
	/** The second coordinate. */
	private final Set ordinate;
	/** The third coordinate. */
	private final Set applicate;
	
	/**
	 * Constructs a new OrderedTriple object.
	 * @param abscissa the first coordinate
	 * @param ordinate the second coordinate
	 * @param applicate  the third coordinate
	 */
	public OrderedTriple(Set abscissa, Set ordinate, Set applicate) {	
		super(new OrderedPair(abscissa, ordinate), applicate);
		this.abscissa = abscissa;
		this.ordinate = ordinate;
		this.applicate = applicate;
	}
	/**
     * Returns whether the specified OrderedTriple is equal to this OrderedTriple.
	 * @param x the specified OrderedPair
	 * @return true if x has the same coordinates as this OrderedPair
	 */
	public boolean equals(OrderedTriple x) {
		boolean a = this.abscissa.equals(x.abscissa);
		boolean b = this.ordinate.equals(x.ordinate);
		boolean c = this.applicate.equals(x.applicate);
		return a && b;
	}
}

File: core/scripts/set/OrderedTriple.js
/**
  * An OrderedTriple is a set that encodes the order of its three coordinates.
  * An ordered triple is written (a, b, c), where a, b, and c are the coordinates.
  *
  * `new OrderedTriple(a,b,c)` is defined as `new OrderedPair(new OrderedPair(a,b), c)`
  * @extends   OrderedPair
  * @param `a` the first coordinate (abscissa)
  * @param `b` the second coordinate (ordinate)
  * @param `c` the third coordinate (applicate)
  */
function OrderedTriple(a, b) {
  this.abscissa  = a || new EmptySet();
  this.ordinate  = b || new Singleton(this.abscissa);
  this.applicate = c || new Singleton(this.ordinate);
  OrderedPair.call(this, new OrderedPair(this.abcissa, this.ordinate), this.applicate);
}
Util.extend(OrderedTriple, OrderedPair);

OrderedTriple.prototype.equals = function equals(x) {
  return this.abscissa.equals(x.abscissa) && this.ordinate.equals(x.ordinate) && this.applicate.equals(x.applicate);
  // fix this, it's wrong. there may be a set that is an ordered triple that is not an instance of
  // the OrderedTriple class
}

OrderedTriple.prototype.toString = function toString() {
  return '( ' + this.abscissa + ',  ' + this.ordinate + ',  ' + this.applicate + ' )';
}
